Keymaps

epatpol opened the pull request:
Signed-off-by: Patrick-Jeffrey Pollo Guilbert <patrick.pollo.guilbert@ericsson.com>

epatpol commented:
Don't mind typos and such, there's lots of stuff to clean up, but as of right now this WIP PR will make it easier to discuss custom kb @akosyakov

svenefftinge commented:
On the client side the keybinding registry should have a `setKeyMap` (instead of individual `override` and `reset`methods). That one basically creates a copy of the default keybindings and overrides them (based on commandid and context) with the given keybindings. Every call to `setKeyMap` will replace that active keybindings array.

Also it would be much simpler to just send the full keybindings from the server, instead of the diffs. (you don't change your keybindings too often).

epatpol commented:
@svenefftinge Right! I wasn't sure if either the service or the registry was the one keeping the default keymaps cache for resetting, and the way you mentionned seems better as I don't have to modify the current keybinding API.

The diffs were for when to reset to a default keybinding (i.e a custom keybinding was deleted). Unless you mean the deletion detection should not happen directly in the server?

svenefftinge commented:
I meant that on every change of the key preference we should send the whole contents and replace it with the current keybindings. So the old overridden state will be replaced.

akosyakov commented on packages/preferences-api/src/common/custom-keybinding-service.ts:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    protected onDidChangePreference(event: KeybindingEvent): void {
+        for (const kbChange of event.changes) {
```
this method should call  `KeybindingRegistry.setKeymap(keymap)` that in its turn should implement: https://github.com/theia-ide/theia/pull/456#issuecomment-327407558

epatpol replied:
In that case shouldn't the RawKeyBinding interface be in the core? Or does keymap here reference something that's already declared in the core?

epatpol commented on packages/preferences/src/node/preference-backend-module.ts:
```diff
@@ -28,10 +29,13 @@ export type UserPreferenceServer = PreferenceServer;
+
+    bind(KeybindingURI).toConstantValue(homeUri.path.join('.theia', 'keybindings.json'));
     bind(JsonPreferenceServer).toSelf();
+    bind(CustomKeybindingServer).toSelf().inSingletonScope();
```
I was planning to do this once but I didn't want to create a whole new package for the keybindings, should I?

akosyakov replied:
let's start with the separate folder in the preferences package which should be easily moveable to a separate extension

epatpol replied:
Ok will do!

epatpol commented:
Maybe we don't even need the service? JSON server sends the whole keymaps json straight to setKeyMap where (if found valid), it overwrites the keybindings, otherwise it leaves the default ones.

epatpol commented:
This is pretty much ready to review. Basic functionality works. You can test by adding a `keybindings.json` under your $HOME/.theia/keybindings.json with the something like 

```json
[
    {
        "command": "quickCommand",
        "keybinding": "F5"
    }
]
```

This should change the keystroke to open the quickCommand with F5. I think it should also work on os x too? If somebody on OS X could test that'd be great.

hexa00 commented:
I'm reviewing that...

hexa00 commented on packages/core/src/common/keys.ts:
```diff
@@ -53,6 +53,30 @@ export class KeyCode {
+            const key = CODE_TO_KEY[keyString];
+            if (Key.isKey(key)) {
+                if (Key.isModifier(key.code)) {
+                    sequence.push(MODIFIERS.filter(item => item.code === key.code)[0].code);
```
can't we just get this direcltty?

hexa00 commented on packages/core/src/common/keys.ts:
```diff
@@ -53,6 +53,30 @@ export class KeyCode {
+            }
+        }
+
+        return new KeyCode(sequence.join('+'));
```
This sounds really weird parse from string then create a keycode from another string with the modifier keys changed...
Could that be directly in Keycode ?
Like Keycode.fromUserConfig() ?

hexa00 commented on packages/custom-keybindings-api/src/common/custom-keybinding-service.ts:
```diff
@@ -0,0 +1,81 @@
+import { FrontendApplicationContribution, FrontendApplication } from '@theia/core/lib/browser';
+import { Disposable, DisposableCollection, CommandRegistry, KeybindingRegistry, ILogger } from '@theia/core/lib/common';
+import { RawKeybinding, KeybindingServer, KeymapChangeEvent } from './keybindings-protocol';
+import * as Ajv from "ajv";
```
this should be `import * as ajv from 'ajv'`

hexa00 commented on packages/custom-keybindings-api/src/common/custom-keybinding-service.ts:
```diff
@@ -0,0 +1,81 @@
+    protected readonly ajv = new Ajv();
+    protected ready = false;
+
+
```
empty lines

hexa00 commented on packages/custom-keybindings-api/src/common/custom-keybinding-service.ts:
```diff
@@ -0,0 +1,81 @@
+
+    protected resolveReady: () => void;
+
+
```
empty lines

hexa00 commented on packages/custom-keybindings-api/src/common/custom-keybinding-service.ts:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    protected onDidChangeKeymap(keymapChangeEvent: KeymapChangeEvent): void {
+        if (this.ready) {
```
I wonder what happens if this is not ready?
You lose the change? Will that break stuff?

epatpol replied:
I think I could remove it, I think it's leftovers from looking at other frontend contributions work, in this case it might not be needed. I was maybe afraid that the keybinding contributions could be loaded after the keymap server reads the keymaps file, whereas they should always be treated first.

hexa00 commented on packages/custom-keybindings-api/src/common/keybindings-protocol.ts:
```diff
@@ -0,0 +1,35 @@
+import { JsonRpcServer } from "@theia/core";
+
+export const keybindingsPath = '/services/keybindings';
+
```
empty line

hexa00 commented on packages/custom-keybindings-api/src/common/keybindings-protocol.ts:
```diff
@@ -0,0 +1,35 @@
+    changes: RawKeybinding[]
+}
+
+// export interface JsonChange {
```
Remove?

hexa00 commented on packages/custom-keybindings/src/browser/keybinding-frontend-module.ts:
```diff
@@ -0,0 +1,20 @@
+import { KeybindingServer, CustomKeybindingService, keybindingsPath } from "../common";
+
+export default new ContainerModule(bind => {
+    // bind(CustomKeybindingService).toSelf().inSingletonScope();
```
remove ?

hexa00 commented on packages/custom-keybindings/src/common/index.ts:
```diff
@@ -0,0 +1,10 @@
+ */
+
+export * from '@theia/custom-keybindings-api';
+
```
empty line

hexa00 commented on packages/custom-keybindings/src/node/custom-keybinding-server.spec.ts:
```diff
@@ -0,0 +1,89 @@
+    return new FileSystemNode();
+}
+
+function createLogger(): ILogger {
```
ditto sinon

hexa00 commented on packages/custom-keybindings/src/node/custom-keybinding-server.ts:
```diff
@@ -0,0 +1,126 @@
+
+    protected fireEvent(event: KeymapChangeEvent) {
+        this.logger.debug(log =>
+            log('onDidChangeKeymap:', event)
```
Please create a child logger as in the terminal module otherwise we have no idea where this log comes from

hexa00 commented on packages/custom-keybindings/src/node/keybinding-backend-module.ts:
```diff
@@ -0,0 +1,42 @@
+    bind(CustomKeybindingServer).toSelf();
+
+    bind(KeybindingServer).to(CustomKeybindingServer);
+
```
we should strip the previous empty lines

epatpol commented on packages/core/src/common/keybinding.ts:
```diff
@@ -204,4 +227,53 @@ export class KeybindingRegistry {
+    setKeymap(rawKeyBindings: RawKeybinding[]) {
+        /* Only assign the default keybindings once, so that they do not get overridden with further iterations */
+        if (!this.defaultKeyBindings || !this.defaultCommands) {
+            this.defaultKeyBindings = Object.assign({}, this.keybindings);
```
From what I've tried if I do this modifying the this.keybindings array by pushing new keybindings on it then also modifies the this.defaultKeyBindings (same reference?). Maybe there's a better way to do this :/

epatpol commented on packages/core/src/common/keys.ts:
```diff
@@ -53,6 +53,30 @@ export class KeyCode {
+
+        const sequence: string[] = [];
+        const keys = keybinding.split('+');
+        for (const keyString of keys) {
```
Yea I thought about that too, especially if we are to use sequences later on

hexa00 replied:
Sequences would use space as a separator so that's an argument against it hehe

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -151,15 +176,18 @@ export class KeybindingRegistry {
      * @param options if `active` is false` then the availability of the command will not be checked. Default is `true`
      */
-    getKeybindingForCommand(commandId: string, options: { active: boolean } = ({ active: true })): Keybinding | undefined {
+    getKeybindingsForCommand(commandId: string, options: { active: boolean } = ({ active: true })): Keybinding[] | undefined {
```
Should return an empty array instead of undefined.

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -140,7 +150,22 @@ export class KeybindingRegistry {
+            const { keyCode, commandId } = bindingToRemove;
+            let bindings = this.keybindings[keyCode.keystroke] || [];
+
+            bindings = bindings.filter(binding => binding !== bindingToRemove);
```
We should have a proper value comparison logic instead of relying on object identity.

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -140,7 +150,22 @@ export class KeybindingRegistry {
+        this.commands[commandId] = commands;
+    }
+
+    unregisterKeybindings(bindingsToRemove: Keybinding[]) {
```
Would be nice to get rid of this modification logic.

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -26,6 +26,13 @@ export interface Keybinding {
     readonly accelerator?: Accelerator;
 }
 
+export interface RawKeybinding {
```
Could we just get rid of KeyCode in the `Keybinding` and use that one instead?

epatpol replied:
If by "that one" you meant the string representation, I think it's doable, but there's going to be refactoring the monaco part. From what I understand having that KeyCode structure makes it easier to rebind the monaco commands to theia commands. Worth investigating for sure, as I think one structure for everything everywhere is better.

svenefftinge replied:
Ok let's do that later, then.

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -26,6 +26,13 @@ export interface Keybinding {
+    command: string;
+    keybinding: string;
+    context?: string;
+    args?: string[];
```
what is args?

akosyakov replied:
default args for the command, the same as VS code

epatpol replied:
I think I should remove that, looking at vscode their custom keybindings structure is simply { command, keybinding and when (context) }. Maybe we should go with a similar approach.

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -204,4 +232,52 @@ export class KeybindingRegistry {
+        }
+        let invalidKeyMap: boolean = false;
+        for (const rawKeyBinding of rawKeyBindings) {
+            if (this.commandRegistry.getCommand(rawKeyBinding.command)) {
```
This code will modify the previous map that is a mixture of defaultbindings and previous keymaps.

epatpol commented:
> Could we go without modifying the keybinding collections, but just replacing them?
So register would contribute to 'defautlKeybinings' directly and the setKeyMap() will replace the keybindings atomically.
I think it would be good to extract a class (e.g. KeybindingIndex) that does the lookup by keybinding (our string representation) and by command. That structure could have a parent relationship, i.e. keymapBindings -> defaultBindings.

I like this approach, alright let me try making a prototype for this

kittaakos commented on packages/core/src/common/keys.ts:
```diff
@@ -148,7 +172,7 @@ const MODIFIERS: Key[] = [];
 
     export function isKey(arg: any): arg is Key {
-        return (<Key>arg).code !== undefined && (<Key>arg).keyCode !== undefined;
+        return arg && (<Key>arg).code !== undefined && (<Key>arg).keyCode !== undefined;
```
I just would like to understand this: do not we have an explicit `undefined`/`null` check enabled?

akosyakov replied:
just of the topic: such code looks dangerous, it can cause side effects by calling code or keyCode, better:
```ts
return !!arg && ('code' in arg) && ('keyCode' in arg);
```

kittaakos replied:
Why is this dangerous? That is the recommended way according to the [TS documentation](https://www.typescriptlang.org/docs/handbook/advanced-types.html), did I miss something? What could be the problem, could you please give us an example, @akosyakov?

akosyakov replied:
- `('code' in arg)` checks only that arg has a property with `code` name
- `(<Key>arg).code !== undefined` calls `code` property and checks whether the result is of `undefined` type - it can be also bogus if there is code property which returns `undefined` value

kittaakos commented on packages/core/src/common/keybinding.ts:
```diff
@@ -115,32 +208,45 @@ export class KeybindingRegistry {
+     *
+     * @param binding
+     */
+    setKeybinding(binding: Keybinding) {
```
Can we call the method `add` or `register` or something like that: just as you did for the documentation.

epatpol replied:
I was thinking maybe register/override (register for contributions) and overriding when coming from the config file?

kittaakos commented on packages/core/src/common/keybinding.ts:
```diff
@@ -53,6 +60,88 @@ export namespace KeybindingContexts {
+    protected readonly defaultCommands: { [commandId: string]: Keybinding[] } = {};
+
+
+    constructor() {
```
I do not see a reason for the empty `ctor` function. So we could get rid of the one in the sub-class too.

svenefftinge commented:
+1 for the recent changes. We should go one step further and only use the string representation outside. So conversion to and from browser keymaps should be done and kept internally.

svenefftinge commented on packages/core/src/browser/common-frontend-contribution.ts:
```diff
@@ -187,49 +187,49 @@ export class CommonFrontendContribution implements MenuContribution, CommandCont
+            registry.registerDefaultKeyBinding({
                 commandId: CommonCommands.CUT.id,
-                keyCode: KeyCode.createKeyCode({ first: Key.KEY_X, modifiers: [Modifier.M1] })
+                keyCode: TheiaKeyCodeUtils.createKeyCode({ first: Key.KEY_X, modifiers: [Modifier.M1] })
```
let's just assign a string in our keybinding syntax here. The name should be changed from keyCode to e.g. `binding`. And get rid of `accelerator`, too.

svenefftinge commented on packages/core/src/browser/quick-open/quick-command-service.ts:
```diff
@@ -67,7 +67,8 @@ export class CommandQuickOpenItem extends QuickOpenItem {
     getKeybinding(): Keybinding | undefined {
-        return this.keybindings.getKeybindingForCommand(this.command.id);
+        const bindings = this.keybindings.getKeybindingsForCommand(this.command.id);
+        return bindings ? bindings[0] : undefined;
```
It think we should introduce `keybindings.getPrimaryKeybinding(commandid)`

epatpol replied:
Good idea, and return the first one I guess?

svenefftinge replied:
for now, yes

svenefftinge commented on packages/core/src/browser/widgets/widget.ts:
```diff
@@ -100,9 +100,9 @@ export function addEventListener<K extends keyof HTMLElementEventMap>(
 
 export function addKeyListener<K extends keyof HTMLElementEventMap>(element: HTMLElement, keybinding: Key, action: () => void, ...additionalEventTypes: K[]): Disposable {
```
here again we should use the string representation rather than `Key`.

epatpol commented:
I've added a file->menu contrib that allows opening up the keymaps file, if unexisting, it will create one with a template to help the user on how to add custom keybindings. Please let me know if the labels/default content of the file is fine.

epatpol commented:
@svenefftinge @akosyakov I've made some changs with the latest commit using a basic user-storage implementation, any comments would be appreciated!

akosyakov commented on packages/keymaps/src/browser/keymaps-frontend-module.ts:
```diff
@@ -21,9 +23,14 @@ export default new ContainerModule(bind => {
 
-    bind(KeymapsServer).toDynamicValue(ctx =>
-        ctx.container.get(WebSocketConnectionProvider).createProxy(keybindingsPath)
+    bind(StorageServer).toDynamicValue(ctx =>
```
`StorageServer` is too generic name, better `UserStorageServer` or `IUserStorageServer` if the former is alread taken

akosyakov commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -0,0 +1,118 @@
+};
+
+// export const keymapsUriString = 'user:/keymaps.json';
+export const keymapsUriString = new URI('').withScheme(UserStorageUri.SCHEME).withPath('keymaps.json').toString();
```
better if it is of `URI` type

I thought you will have something like:
```ts
export const keymapsUri = UserStorageUri.toUri('keymaps.json')
```

akosyakov commented on packages/keymaps/src/browser/user-storage-resource.ts:
```diff
@@ -0,0 +1,75 @@
+        this.toDispose.push(this.watcher.onFilesChanged(changes => {
+            for (const changedUri of changes.uris) {
+                const uri = new URI(changedUri)
+                if (uri.toString() === this._uri) {
```
`changedUri === this._uri` should be enough, no need conversions

akosyakov commented on packages/keymaps/src/browser/user-storage-resource.ts:
```diff
@@ -0,0 +1,75 @@
+
+    protected readonly onDidChangeContentsEmitter = new Emitter<void>();
+    protected readonly toDispose = new DisposableCollection();
+    private _uri: string;
```
readonly?

akosyakov commented on packages/keymaps/src/common/user-storage-protocol.ts:
```diff
@@ -0,0 +1,23 @@
+import { JsonRpcServer } from "@theia/core";
+export const userStoragePath = '/services/storage';
+export const StorageServer = Symbol("StorageServer");
+export interface StorageServer extends JsonRpcServer<UserStorageClient> {
```
`UserStorageServer` or `IUserStorageServer`, the same about the symbol name

akosyakov commented on packages/keymaps/src/common/user-storage-uri.ts:
```diff
@@ -0,0 +1,28 @@
+
+export namespace UserStorageUri {
+
+    export const SCHEME = 'user';
```
`user_storage` would be more clear

akosyakov commented on packages/keymaps/src/common/user-storage-uri.ts:
```diff
@@ -0,0 +1,28 @@
+     */
+    export function create(fsPath: string) {
+        const fsUri = new URI(fsPath);
+        return new URI('').withScheme(SCHEME).withPath(fsUri.displayName);
```
Why `fsUri.displayName`?

epatpol replied:
Should it be fsUri.path.base? If I use fsUri.path.toString() I end up with user_storage:/home/user/.theia/.keymaps.json. Is there an easy way to strip the relative path from theia user dir (i.e not just the filename)? This way */home/user/.theia*/folder/keymaps.json we could extract */folder/keymaps.json*

akosyakov commented on packages/keymaps/src/browser/user-storage-resource.ts:
```diff
@@ -0,0 +1,75 @@
+
+        this._uri = uri.toString();
+        this.toDispose.push(this.onDidChangeContentsEmitter);
+        this.toDispose.push(this.watcher.onFilesChanged(changes => {
```
Do we really need `UserStorageWatcher`?  Why not install a client on `UserStorageServer` here?

epatpol replied:
Because I thought it overrides the last client registered, doesn't it?

akosyakov commented on packages/keymaps/src/common/user-storage-watcher.ts:
```diff
@@ -0,0 +1,48 @@
+
+
+@injectable()
+export class UserStorageWatcher implements Disposable {
```
I thought that watchers are redundant since we can call `server.setClient` anywhere.

epatpol replied:
I tried that before and weirdly enough only events coming from outside of theia worked. I thought it was the server not being a singleton but even with this modifying from within theia always failed. Maybe something got disposed too quickly?

epatpol commented:
@svenefftinge sorry I missed the last comment! This one is ready for review. There are 1 or 2 things I'm not sure yet, i.e should user storage be in its own package (if so, I ended up with a keymaps package with only one classes so...).

Thanks!

epatpol commented:
@sven @akosyakov Do you think there should be more stuff changed in this PR to make it OK in relation with the multiple frontends/single backend? As the keymaps should only be in the user's home directory (kinda like vscode) and not in workspace, I don't think this should be affected.

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -108,65 +195,77 @@ export class KeybindingRegistry {
-        const commands = this.commands[commandId] || [];
-        commands.push(binding);
-        this.commands[commandId] = bindings;
+        this.keymapsIndex.setParentKeyBinding(binding);
```
Should do `this.defaultKeybindingIndex.setKeyBinding(binding)` instead.

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -53,6 +59,88 @@ export namespace KeybindingContexts {
+        return false;
+    }
+
+    setParentKeyBinding(keybinding: Keybinding): void {
```
We should just keep a handle to the parent and remove these two 'parent'-methods

epatpol replied:
This is basically just setting the keybindings from the contributions, you mean maybe the name isn't correct? Basically there's already a handle to the parent here
```
constructor(private parent?: KeybindingIndex) 
```

Unless you mean something else?

svenefftinge replied:
I meant the only client to these methods is below (see my comment). And if we instead use the defaultKeybindingIndex we don't need these methods at all.

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -108,65 +195,77 @@ export class KeybindingRegistry {
+     *
+     * @param binding
+     */
+    overrideKeyBinding(binding: Keybinding) {
```
Looks like this method would check both indexes, but should not since, I want to 'override' the parent index.

svenefftinge replied:
Couldn't this override logic go into `keymapsIndex.setKeyBinding`?

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -187,11 +286,21 @@ export class KeybindingRegistry {
         if (!binding) {
             return;
         }
+        const bindingsForThisCommand = this.getKeybindingsForCommand(binding.commandId);
```
It is not clear to me, why you do these checks here. We already have a binding, couldn't we just execute it?

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -203,4 +312,42 @@ export class KeybindingRegistry {
         }
     }
 
+    setKeymap(rawKeyBindings: RawKeybinding[]) {
```
Couldn't we just create a new keymapsIndex using the given rawKeyBindings?
The check for existing commands and context should just filter down the given `rawKeyBindings` and log errors.
Afterwards we should just do something like
```
   this.keymapsIndex = new KeybindingIndex(this.defaultKeybindingIndex, filteredKeyBindings)
```

epatpol commented:
@svenefftinge I rebased/squashed most commits as the history for this PR didn't make sense anymore, as some things were being added in the first commit, and then editted/removed in later commits and so on. It's bigger commits but it makes more sense as a whole.

svenefftinge commented:
the tests are failing, could you check that?

svenefftinge commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -0,0 +1,114 @@
+    protected readonly toDispose = new DisposableCollection();
+    protected readonly ajv = new ajv();
+    protected ready = false;
+    protected keymapsResource: UserStorageResource;
```
No need to expect a `UserStorageResource`, the general interface `Resource` should be enough.

svenefftinge commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -0,0 +1,114 @@
+                this.setKeymap(keybindings);
+            }
+        } catch (error) {
+            this.keyBindingRegistry.resetKeybindings();
```
There were keybindings registered that worked. Now the user changed the keybindings and broke the JSON and as a result we reset the working keybindings? I don't think that should happen.

epatpol replied:
As stated before, I did it this way because VScode does it like that. To me it makes sense to provide zero custom keybindings if your keybindings file is invalid, this way I really know something is wrong with my config.

svenefftinge replied:
ok, let's go with this behavior and see how it feels in practice.

svenefftinge commented on packages/core/src/common/keybinding.ts:
```diff
@@ -248,4 +316,42 @@ export class KeybindingRegistry {
+                    });
+
+                } else {
+                    this.resetKeybindings();
```
If you want to restore, then please to the previous state and not to 'no keymaps'.

epatpol replied:
But if the json file is invalid, I'd rather have the default keybindings than the previous keybindings no? I went this way as VScode does it like this too. Wdyt?

svenefftinge commented on packages/cpp/src/browser/cpp-keybinding.ts:
```diff
@@ -31,15 +31,15 @@ export class CppKeybindingContribution implements KeybindingContribution {
                 commandId: 'switch_source_header',
                 context: this.cppKeybindingContext,
-                keyCode: KeyCode.createKeyCode({ first: Key.KEY_O, modifiers: [Modifier.M3] })
+                keyCode: TheiaKeyCodeUtils.createKeyCode({ first: Key.KEY_O, modifiers: [Modifier.M3] })
```
I wonder if renaming `registerKeybinding` -> `registerDefaultKeybindings` is really an improvement, now that keymaps can only be set at once. Also `KeyCode`->`TheiaKeyCodeUtils` is not better IMHO. The whole PR would be much smaller if your reverse those renames.

epatpol replied:
Ok for the `registerKeybinding`. I think I changed it to TheiaKeyCodeUtils as I ended up with a lot of helper functions in the KeyCode class to make the easy naming keys work... I can revert it back also.

svenefftinge commented on packages/keymaps/src/browser/keymaps-frontend-contribution.ts:
```diff
@@ -0,0 +1,53 @@
+export namespace KeymapsCommands {
+    export const OPEN_KEYMAPS: Command = {
+        id: 'keymaps:open',
+        label: 'Open Keyboard Shortcuts'
```
`Open Keyboard Shortcuts` would open an empty file. I think that is a bit surprising for users.
Should we maybe generate a json based on the defaults on first access and then use some diff tool to find the differences?
Also the we use auto-save by default, so there will be frequent contentChange signals on the resource with broken syntax.

epatpol replied:
Yea we just discussed this, I think maybe for now opening two editors side by side would be nice. The one on the left with the default keybindings provided to the registry (readonly) and the one on the right the custom keymaps.json. A sort of diff editor would be nice, but I don't want it to show all the differences like a default diff editor would do... 

I'm not sure how to resolve the frequence contentChange signals problem, isn't that a common problem with using that setting anyway?

epatpol replied:
Also can we do a separate PR for opening the default shortcuts?

svenefftinge replied:
> Also can we do a separate PR for opening the default shortcuts?

Yes, let's do that

epatpol commented:
I reverted the namings of both KeyCode and registerKeybinding, now let's just hope the tests pass.

svenefftinge commented on packages/core/src/common/keys.ts:
```diff
@@ -90,6 +129,80 @@ export class KeyCode {
         }
     }
 
+    // public static createKeyCode(event: KeyboardEvent | Keystroke): KeyCode {
```
please delete

svenefftinge commented on packages/keymaps/README.md:
```diff
@@ -0,0 +1,6 @@
+# Theia - Keymaps Extension
+
+See [here](https://github.com/theia-ide/theia) for a detailed documentation.
```
a short description would be good.

epatpol commented:
@svenefftinge tell me if you have time to check this one again before we merge it. We can open new issues after on what needs to be done to further improve it.

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -72,6 +77,78 @@ export namespace KeybindingContexts {
+    get keybindings(): Keybinding[] {
+        return this.bindings;
+    }
+}
```
Note by default extensions register to the default scope so you could have a 
`setKeybinding(binding: Keybinding)` too

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -170,7 +237,7 @@ export class KeybindingRegistry {
      */
     getKeybindingsForCommand(commandId: string): Keybinding[] {
-        return this.commands[commandId] || [];
+        return this.keymapsIndex.getKeybindingsForCommand(commandId) || [];
```
I think this is weird... we should merge the keymapIndex into the keybinding registry.. there's no need IMO to have 2 entities .. the KeybindingRegistry should be smart enough on it's own to understand scopes.

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -179,8 +246,8 @@ export class KeybindingRegistry {
      */
-    getKeybindingsForKeyCode(keyCode: KeyCode): Keybinding[] {
-        const bindings = this.keybindings[keyCode.keystroke] || [];
+    getKeybindingForKeyCode(keyCode: KeyCode): Keybinding[] {
```
why is this renamed singular but still return an array of keybindings?

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -229,7 +296,7 @@ export class KeybindingRegistry {
 
         const keyCode = KeyCode.createKeyCode(event);
-        const bindings = this.getKeybindingsForKeyCode(keyCode);
+        const bindings = this.getKeybindingForKeyCode(keyCode);
```
weird here too.. singular goes to plural...

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -267,4 +334,42 @@ export class KeybindingRegistry {
+                // TODO check if valid context? valid keybinding?
+                // Create Keycode from string?
+                // Create new Keybinding and assign it
+                // If one keybinding is wrong, reset to the default ones.
```
Remove todo ? context is not in question any context is ok right?

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -267,4 +334,42 @@ export class KeybindingRegistry {
+                // Create new Keybinding and assign it
+                // If one keybinding is wrong, reset to the default ones.
+
+                const code = KeyCode.parseKeystroke(rawKeyBinding.keybinding);
```
This should be KeyCode.parse(string) since it returns a KeyCode... the documentation is also wrong as:

```typescript
    /**
     * Parses a Keystroke object from a string.
     * @param keybinding String representation of a keybinding
     */
    public static parseKeystroke(keybinding: string): KeyCode | undefined {
```
This creates a KeyCode.. there's no such thing as a Keystroke object

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -267,4 +334,42 @@ export class KeybindingRegistry {
+                return;
+            }
+        }
+        this.keymapsIndex = new KeybindingIndex(this.defaultKeybindingIndex, customBindings);
```
This would just become this.keymaps[scope] = customBindings

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -267,4 +334,42 @@ export class KeybindingRegistry {
         return commandId === KeybindingRegistry.PASSTHROUGH_PSEUDO_COMMAND;
     }
+
+    setKeymap(rawKeyBindings: RawKeybinding[]) {
```
setKeymap(scope, rawKeybidings)

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -267,4 +334,42 @@ export class KeybindingRegistry {
+
+    resetKeybindings() {
+        this.keymapsIndex = new KeybindingIndex(this.defaultKeybindingIndex);
+    }
```
```typescript
resetKeybindings(scope) {

  this.keymaps[scope] =  []
}
```

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -72,6 +77,78 @@ export namespace KeybindingContexts {
+        return false;
+    }
+
+    setKeyBinding(keybinding: Keybinding): void {
```
This should support a Keybinding | string
and if it's a string use the easy keys

epatpol replied:
I think it's not going to be called directly as the string will be parsed before that, as you really need the whole struct to set a keybinding (command id, context, keystroke) etc. Not just the keystroke I think

hexa00 commented on packages/core/src/common/keys.ts:
```diff
@@ -53,6 +53,45 @@ export class KeyCode {
+     * Parses a Keystroke object from a string.
+     * @param keybinding String representation of a keybinding
+     */
+    public static parseKeystroke(keybinding: string): KeyCode | undefined {
```
like said before.. this should just be parse...

hexa00 commented on packages/core/src/common/keys.ts:
```diff
@@ -114,6 +153,19 @@ export class KeyCode {
+            return keycode1.alt === keycode2.alt &&
+                keycode1.ctrl === keycode2.ctrl &&
+                keycode1.meta === keycode2.meta &&
+                keycode1.shift === keycode2.shift && keycode1.key.keyCode === keycode2.key.keyCode &&
```
continue using 1 operation per line ?

hexa00 commented on packages/core/src/common/keys.ts:
```diff
@@ -114,6 +153,19 @@ export class KeyCode {
+                keycode1.ctrl === keycode2.ctrl &&
+                keycode1.meta === keycode2.meta &&
+                keycode1.shift === keycode2.shift && keycode1.key.keyCode === keycode2.key.keyCode &&
+                keycode1.key.code === keycode2.key.code;
```
We should compare keycode1.key with keycode2.key directly you could use JSON.stringify() for that since the object have the same properties order and are simple.

epatpol replied:
So just to be clear I should use JSON.stringify(object1) === JSON.stringify(object2)? That makes it much cleaner I agree.

hexa00 replied:
Yes it's just that if you add an element in object1 you won't break your compare here.

hexa00 commented on packages/keymaps/README.md:
```diff
@@ -0,0 +1,21 @@
+        "command": "quickCommand",
+        "keybinding": "ctrl+shift+f4"
+    }
+]
```
I think we need to define the complete list of 'easy keys' here.

epatpol replied:
Would a link to `keys.ts` with the correct line number of the map be enough as a reference for the readme?

hexa00 replied:
I think since these are static and may be geared at end users it would be better to copy them in the readme.

hexa00 replied:
See how they specific this here: https://craig.is/killing/mice as example?

epatpol replied:
With that example that makes sense, I'll add it to the readme.

hexa00 commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -0,0 +1,116 @@
+    "additionalProperties": false
+};
+
+export const keymapsUri: URI = new URI('keymaps.json').withScheme(UserStorageUri.SCHEME);
```
does they mean that any file named keymaps.json anywhere will open with that?

What if I'm developing an app that has a file called like this...?

epatpol replied:
Because the scheme is "user_storage" this makes it so that the user storage service is in charge of resolving it (and checking if it exists in the user storage location i.e user home or something like this).

hexa00 replied:
OK :)

hexa00 commented:
Also thinking about this is there an issue if I want to have a keybinding like `ctr +` ?  
Since it splits on `+` I can't write `Ctrl++` right?

epatpol commented on packages/core/src/common/keybinding.ts:
```diff
@@ -72,6 +77,78 @@ export namespace KeybindingContexts {
+    getKeybindingsForCommand(command: string): Keybinding[] {
+        const result: Keybinding[] = [];
+
+        if (this.parent) {
```
I'm trying to get the correct use case around my head, if we ask for keybindings for a command `testCommand` and we see that the workspace scope already returns 1 (or 2 or 3) keybinding, shouldn't we just stop and return those already? They already "override" the ones above for the same commandId, or am I missing something here?

hexa00 replied:
For reference as discussed offline, yes this would be important if at some time we want to know what is being shadowed by what...

hexa00 commented:
Actually we could solve that by adding `plus` as a special character like they do in mousetrap see: https://craig.is/killing/mice 
Other special keys are backspace, tab, enter, return, capslock, esc, escape, space, pageup, pagedown, end, home, left, up, right, down, ins, del, and plus.

epatpol commented:
I guess the ideas of the keys in the readme makes sense too for the special characters.

epatpol commented:
@hexa00 I've pushed a new patch with most or your changes done (I still need to update the readme properly, add a test for the latest bugs I found, and modify the equals to use jsonify or something). At first I wanted to add a second commit on top of the first one so you could see the changes I made more easily, but I ended up rebasing everything and squashing it with the first commit (sorry for that). 

Anyway it works very well and with that I think I could easily add workspace keybindings. Let me know what you think!

epatpol commented on packages/core/src/common/keybinding.ts:
```diff
@@ -170,7 +171,20 @@ export class KeybindingRegistry {
+                return result;
+            }
+        }
+        return result;
```
@hexa00 Not even used at this moment, but should I go further and return all keybindings (shadowed or not) associated with a command id? I know this might be useful in later iterations for the editor and whatnot, but maybe we could add that further down the line?

epatpol commented on packages/core/src/common/keybinding.ts:
```diff
@@ -113,9 +125,8 @@ export class KeybindingContextRegistry {
 @injectable()
 export class KeybindingRegistry {
 
+    private keymaps: Keybinding[][] = [[], [], [], []];
```
Is there a better way to initialize this? I'd like for it to be bound by the amount of constants in the enum above, without having to do it in the constructor...

hexa00 replied:
keybinding[][keybindingscope.end] ?

epatpol replied:
Still makes it so that this.keymaps[0] is undefined no?

epatpol replied:
The other option I had is this in the constructor... 
```
        this.keymaps = [];
        Object.keys(KeybindingScopes).forEach(key => this.keymaps.push([]));
```

hexa00 replied:
That wont work since:

```
> Object.keys(KeybindingScopes)
[ '0', '1', '2', '3', 'DEFAULT', 'USER', 'WORKSPACE', 'END' ]
```

Better of using something like 
`for let i = 0; i < KeybingScopes.END; i++) { this.keymaps.push([])}`

epatpol replied:
Oops you're right!

epatpol commented:
@hexa00 I've changed the code a bit to support all the strings mousetrap support (and also 'plus') and added the appropriate documentation. I've added a test to make sure several level scopes also work. Let me know what you think when you can.

@svenefftinge I've added a dev-dependency to jsdom-global which allows us to easily make tests run in mocha (node) even though they have references to KeyboardEvent (DOM api). Otherwise I couldn't test KeyCode.createKeycode(KeyboardEvent | keystroke) as it had a reference to the dom api. Let me know what you think. I feel like my use case was simple but it could be used in other places in core to unit test frontend stuff. I was also wondering if I put it at the correct place i.e in core/package.json and not /package.json (the root one).

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -267,4 +303,37 @@ export class KeybindingRegistry {
+        this.keymaps[scope] = customBindings;
+    }
+
+    resetKeybindings(scope: number) {
```
ForScope? seems like you could have a reset all keybindings function.. that would be called resetKeybindings ?

hexa00 commented on packages/core/src/common/keybinding.ts:
```diff
@@ -267,4 +303,37 @@ export class KeybindingRegistry {
+                    return;
+                }
+            } else {
+                this.resetKeybindings(scope);
```
Add user warning here?

hexa00 replied:
Also this only means to that the user tried to define a wrong command... is it worth resetting in that case ? 
I understand you want to do that if the whole keymap fails to parse but for individual bindings I think we should just warn the user ?

epatpol replied:
I agree, I'll log an error for now.

hexa00 commented on packages/core/src/common/keys.ts:
```diff
@@ -140,15 +196,128 @@ export enum Modifier {
+    export const KEY_X: EasyKey = { keyCode: 88, easyString: 'x' };
+    export const KEY_Y: EasyKey = { keyCode: 89, easyString: 'y' };
+    export const KEY_Z: EasyKey = { keyCode: 90, easyString: 'z' };
+
```
empty line

hexa00 commented on packages/core/src/common/keys.ts:
```diff
@@ -140,15 +196,128 @@ export enum Modifier {
+    export const KEY_Z: EasyKey = { keyCode: 90, easyString: 'z' };
+
+    export const PLUS: EasyKey = { keyCode: 107, easyString: 'plus' };
+
```
empty line

hexa00 commented on packages/core/src/common/keys.ts:
```diff
@@ -53,6 +53,48 @@ export class KeyCode {
+                        sequence.push(`${Modifier.M1}`);
+                    } else if (key.keyCode === EasyKey.SHIFT.keyCode) {
+                        sequence.push(`${Modifier.M2}`);
+
```
empty line

hexa00 commented on packages/core/src/common/keys.ts:
```diff
@@ -53,6 +53,48 @@ export class KeyCode {
+            }
+            const key = EASY_TO_KEY[keyString];
+
+            if (keyString === 'meta') {
```
Could meta just be an easy key too ?

epatpol replied:
Actually for meta/command it's a bit of a workaround. I can't map it to easy key as the keyCode [is browser dependent](https://stackoverflow.com/a/3922353)... So instead theia already has the logic on keydown to check if the meta key was pressed in `createKeyCode` with the help of the `Modifier` map in `keys.ts`.

I think it's easier to just handle it as a separate case here. I could add `cmd` to the special aliases though. 

With the way it works right now, that means a mac user could bind commands using `cmd`, `command` or `meta` and they would work on mac os, but if he switches to Windows/Linux, they would still work (with ctrl instead of command). I'm not sure I answered correctly, so let me know if that's what you meant.

hexa00 replied:
OK I think it just caught my attention.. from my c/cpp that const key = EASY_TO_KEY[keyString]; is called with an invalid keyString.

But I guess it's OK.

epatpol commented on packages/core/src/common/keybinding.ts:
```diff
@@ -180,11 +194,33 @@ export class KeybindingRegistry {
+            this.keymaps[scope].forEach(binding => {
+                if (KeyCode.equals(binding.keyCode, keyCode)) {
+
+                    const _isMoreSpecificBinding: (genericScope: number, matchingBinding: Keybinding) => boolean = (genericScope: number, matchingBinding: Keybinding) => {
```
I got inspired by something I saw passing in a [PR](https://github.com/theia-ide/theia/pull/906/files#diff-7635ea30665061a97be3d5fa78ec6d93R115) a while ago, and I felt like the scope of this function was only to be used there. I don't mind renaming/extracting it at the class level though.

epatpol commented on packages/core/src/common/keybinding.ts:
```diff
@@ -180,11 +194,33 @@ export class KeybindingRegistry {
+                            return false;
+                        }
+
+                        const specificBindings = this.keymaps[genericScope + 1].filter(specificBinding => matchingBinding.commandId === specificBinding.commandId);
```
Ah now "extracting it makes more sense! `getKeybindingsForCommand` would also use it, kinda like how `isShadowed` was before with the index.

epatpol replied:
Also I think you meant  `let nextScope = ++scope;` otherwise it checks for the same scope (instead of checking for the more specific scope). I just pushed a new patch with the changes (and the correct prefix increment).

akosyakov commented on examples/browser/package.json:
```diff
@@ -25,15 +25,16 @@
     "watch": "yarn build --watch",
     "start": "theia start",
-    "start:debug": "yarn start --logLevel=debug",
+    "start:debug": "yarn start --loglevel=debug",
```
it is `logLevel`, not in the lower case: https://github.com/theia-ide/theia/blob/6cfdece584c4615dc0b7b3c4a6b33faddcb91fab/packages/core/src/node/bunyan-logger-server.ts#L21

epatpol replied:
I have no idea how I ended up modifying that (and I didn't notice it when doing `git add --patch`!

akosyakov commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -0,0 +1,116 @@
+        @inject(KeybindingRegistry) protected readonly keyBindingRegistry: KeybindingRegistry,
+        @inject(ILogger) protected readonly logger: ILogger
+    ) {
+        this.resourceProvider(keymapsUri).then(resource => {
```
it cannot be undefined, it will reject

akosyakov commented on packages/keymaps/package.json:
```diff
@@ -0,0 +1,52 @@
+    "@theia/core": "^0.3.0",
+    "@theia/workspace": "^0.3.0",
+    "@theia/userstorage": "^0.3.0",
+    "jsonc-parser": "^1.0.0",
```
I cannot see you use it anywhere

akosyakov commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -0,0 +1,116 @@
+    protected parseKeybindings(content: string) {
+        let keybindings: any;
+        try {
+            keybindings = JSON.parse(content);
```
not `jsonc-parser`?

epatpol replied:
Well spotted, I had it removed when we had webpack problems and didn't put it back, I'll fix it asap.

akosyakov commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+
+    protected setKeymap(keybindings: any | undefined) {
```
Can it be undefined? if so why don't you check it?

epatpol replied:
Cannot be undefined, so I removed it!

akosyakov commented on packages/core/src/common/keybinding.ts:
```diff
@@ -267,4 +330,50 @@ export class KeybindingRegistry {
         return commandId === KeybindingRegistry.PASSTHROUGH_PSEUDO_COMMAND;
     }
+
+    setKeymap(scope: number, rawKeyBindings: RawKeybinding[]) {
```
Should not `scope` be of a type `KeybindingScopes`?

akosyakov commented on packages/core/src/common/keybinding.ts:
```diff
@@ -26,6 +26,13 @@ export interface Keybinding {
     readonly accelerator?: Accelerator;
 }
 
+export enum KeybindingScopes {
```
It should be `KeybindingScopes`, not plural

epatpol replied:
You mean `KeybindingScope` ?

akosyakov commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -0,0 +1,116 @@
+                });
+
+            } else {
+                this.logger.error('No resource exists for: ', keymapsUri);
```
should go to catch

epatpol replied:
Or just remove the error now that you mentioned it above. With the user-storage-service correctly implemented (fixed in #963) it now returns something even if the resource doesn't exist.

akosyakov commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -0,0 +1,116 @@
+export const keymapsUri: URI = new URI('keymaps.json').withScheme(UserStorageUri.SCHEME);
+
+@injectable()
+export class KeymapsService implements Disposable, FrontendApplicationContribution {
```
What about the workspace scope? Is it out of the scope of this PR?

epatpol replied:
Yea I was thinking about it after (in another PR). Maybe we could extract a generic class from keymaps-service that could be implemented by both `KeymapsServiceUser` and `KeymapsServiceWorkspace`.
